ingest/anndata_.py

```diff
@@ -127,3 +135,119 @@ def delocalize_extracted_files(file_path, study_file_id, files_to_delocalize):
+        """Return True if names are unique, else false
+        Expected name_types: ["Feature", "Obs"]
+        """
+        # check feature_name and obs names, feature_id logic not included
+        # TODO (SCP-5105) non-happy path - add feature_id assessment
+        if len(names) == len(names.unique()):
+            return True
+        else:
+            dups = list_duplicates(names)
+            features_for_msg = 2
+            end = features_for_msg if len(dups) > features_for_msg else len(dups)
+            dup_list = dups[:end]
+            dup_string = " ".join(dup_list)
+
+            msg = (
+                f"{name_type} names must be unique within a file. "
+                f"{len(dups)} duplicates found, including: {dup_string}"
+            )
+            GeneExpression.log_for_mixpanel(
+                "error", "content:duplicate:values-within-file", msg
+            )
+            raise ValueError(msg)
+
+    def transform(self):
+        """Transforms matrix into gene data model."""
+        # initialize settings for mock data loads in tests
+        self.test_models = None
+        self.models_processed = 0
+
+        # derive file name from file path
+        file_name = os.path.basename(self.file_path)
+        start_time = datetime.datetime.now()
+        GeneExpression.dev_logger.info("Starting run at " + str(start_time))
+        num_processed = 0
+        gene_models = []
+        data_arrays = []
+        for all_cell_model in GeneExpression.create_data_arrays(
+            name=f"{file_name} Cells",
+            array_type="cells",
+            values=self.adata.obs.index.tolist(),
+            linear_data_type="Study",
+            linear_data_id=self.study_file_id,
+            **self.data_array_kwargs,
+        ):
+            data_arrays.append(all_cell_model)
+
+        # ASSUMPTION all_cell_model same for raw_count and processed_expression
+        # TODO (SCP-5103): if raw counts is indicated check that .raw slot is populated
+
+        # Iterate over feature names (for happy path)
+        for feature in self.adata.var_names.tolist():
+            print(f"processing feature: {feature}")
+            feature_expression_series = sc.get.obs_df(self.adata, keys=feature)
+            if feature_expression_series.hasnans:
+                msg = (
+                    f'Expected numeric expression score - '
+                    f'expression data has NaN values for feature "{feature}"'
+                )
+                GeneExpression.log_for_mixpanel(
+                    "error", "content:type:not-numeric", msg
+                )
+                raise ValueError(msg)
+            # capture sparse (only non zero values and their cell IDs)
+            # check mtx.py for all zero gene handling
+            filtered_expression_series = feature_expression_series[
+                feature_expression_series.values > 0
+            ]
+
+            exp_cells = filtered_expression_series.index.tolist()
```

Comment:
How do we think this will perform for large (> 10GB) expression matrices?  Especially if they're particularly dense?  Do you think we run the risk of running out of RAM?